src/common/libhostlist/hostlist.h

```diff
@@ -0,0 +1,147 @@
+/************************************************************\
+ * Copyright 2020 Lawrence Livermore National Security, LLC
+ * (c.f. AUTHORS, NOTICE.LLNS, COPYING)
+ *
+ * This file is part of the Flux resource manager framework.
+ * For details, see https://github.com/flux-framework.
+ *
+ * SPDX-License-Identifier: LGPL-3.0
+\************************************************************/
+
+/* Functions for encoding/decoding and manipulating RFC 29 Hostlists */
+
+#ifndef FLUX_HOSTLIST_H
+#define FLUX_HOSTLIST_H
+
+struct hostlist *hostlist_create (void);
+
+void hostlist_destroy (struct hostlist *hl);
+
+/*  Decode string 's' in RFC 29 Hostlist Format.
+ *
+ *  Returns hostlist on success or NULL on failure with errno set.
+ */
+struct hostlist *hostlist_decode (const char *s);
+
+/*
+ *  Encode a hostlist to a string in RFC 29 Hostlist Format
+ */
+char *hostlist_encode (struct hostlist *hl);
+
+/*
+ *  Copy a hostlist
+ */
+struct hostlist *hostlist_copy (const struct hostlist *hl);
+
+/*  Add a string representation of a hostlist onto the tail of the
+ *   hostlist 'hl'.
+ *  Returns the total number of hosts appended to 'hl', or -1 on failure.
+ */
+int hostlist_append (struct hostlist *hl, const char *hosts);
+
+/*
+ *  Append hostlist 'hl2' onto 'hl1'.
+ *
+ *  Returns number of hosts appended for success, -1 for failure.
+ *
+ */
+int hostlist_append_list (struct hostlist *hl1, struct hostlist *hl2);
+
+/*
+ *  Return the last host in the hostlist 'hl' or NULL if hostlist is empty
+ *   or there was an error allocating memory.
+ *   The host is removed from the hostlist.
+ *
+ *  Note: Caller is responsible for freeing the returned memory.
+ */
+char * hostlist_pop (struct hostlist * hl);
```

Comment:
nit organizational thing: shouldn't  `hostlist_shift` be listed first b/c it returns first host, while pop returns the last and should be listed second?  I initially thought the "last host" in this comment was a typo.

Reply:
Interesting, I never assumed anything about the order that functions are listed in the header file. Is there some standard or idiom of which I'm not aware?

Another thought, now that `hostlist_first()` sets cursor, `hostlist_shift()` could be replaced with `hostlist_first(hl); hostlist_remove_current (hl);`. If we add a `hostlist_last()` then similarly `hostlist_pop()` could be removed.

Reply:
> Interesting, I never assumed anything about the order that functions are listed in the header file. Is there some standard or idiom of which I'm not aware?

No, just a generality.  For example, I would always think `list_first()` should be listed before `list_next()` and `list_head()` before `list_tail()`.

Perhaps part of this was the use of the word `pop()` and the fact this data structure has the word "list" in it.  So I may have subconsciously thought `first host` as well.

> Another thought, now that hostlist_first() sets cursor, hostlist_shift() could be replaced with hostlist_first(hl); hostlist_remove_current (hl);. If we add a hostlist_last() then similarly hostlist_pop() could be removed.

I think this is a good idea.  I think instinctively I thought there should be a "push()" function to go along with the "pop()" function.

Reply:
Yeah, I renamed `push` to append for clarity, so it makes sense to try to get rid of `pop` as well.